src/transformer_deploy/utils/code_utils.py

```diff
@@ -0,0 +1,58 @@
+#  Copyright 2022, Lefebvre Dalloz Services
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import importlib
+import inspect
+from typing import Any, Dict
+
+
+def update_module(module_name: str, function: Any, new_function_name: str, modifications: Dict[str, str]):
+    """
+    This function helps updating a module given the function name and the modifications to be done on this function
+    Once you use update_module(), you just need to override the function with its new version using the new function
+    name. If you want to replace the whole code function you can pass * in src code as a key.
+    :param module_name: the module to be updated
+    :param function: the function to be updated in the given module
+    :param modifications: a dictionary containing all the modifications to be done, keys are the source/original code
+    and values are the new code to be used to replace source code
+    Example:
+    if you're updating the forward function in T5Attention transformers
+    `transformers.models.t5.modeling_t5.T5Attention.forward` and using `updatedForward` as new function name, you can
+    do:
+        >>> import transformers
+        >>> update_module(module_name="transformers.models.t5.modeling_t5",
+        >>>               function=transformers.models.t5.modeling_t5.T5Attention.forward ,
+        >>>               new_function_name="updatedForward",
+        >>>               modifications=dict("*", "return True"))
+        >>> transformers.models.t5.modeling_t5.T5Attention.forward = transformers.models.t5.modeling_t5.updatedForward
+    """
+    model_module = importlib.import_module(name=module_name)
+    function_code = inspect.getsource(function)
+    for src_code, new_code in modifications.items():
+        try:
+            def_code_idx = function_code.find(
+                "):\n",
+            )
+            function_code = function_code.replace("):\n", "\n):\n")
+            indent_def_len = function_code.find("def")
+            indent = "".join(" " for _ in range(indent_def_len))
```

Comment:
Can you please provide some info about this part of the code? What it does? Why is it needed?
At first view, it seems to be be very specific to the file we are manipulating

I would have expected a simple `replace`

Reply:
No, in fact it is not specific to this file, the idea here is to be aware of the indentation so that we can always add the right indentation. i'll add more comments in the function to explain the logic.

Reply:
Why not using `cleandoc`?
https://docs.python.org/3/library/inspect.html#inspect.cleandoc

Reply:
I will test it